pkg/s3/client.go

```diff
@@ -0,0 +1,101 @@
+package s3
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	sync "github.com/bacalhau-project/golang-mutex-tracer"
+)
+
+type ClientWrapper struct {
+	S3         *s3.Client
+	Downloader *manager.Downloader
+	Uploader   *manager.Uploader
+	Endpoint   string
+	Region     string
+}
+
+type ClientProviderParams struct {
+	AWSConfig aws.Config
+}
+
+type ClientProvider struct {
+	awsConfig aws.Config
+	clients   map[string]*ClientWrapper
+	mu        sync.RWMutex
+}
+
+func NewClientProvider(params ClientProviderParams) *ClientProvider {
```

Comment:
I have been following the practice of passing structs wrapping params instead of directly, which minimizes the changes when want to pass additional params, and will play nicely once we start using Fx for dependency injection

Reply:
In that case my preference would be to use the [options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) for optional fields on the struct (if options are absent a sane default is used) and to accept required fields as params if when they are not optional, but no need to block the merge on this. Motivation here is to avoid passing methods half populate structs as arguments.